iocore/hostdb/HostDB.cc

```diff
@@ -77,6 +77,46 @@ HostDBCache hostDB;
 
 void ParseHostFile(const char *path, unsigned int interval);
 
+//////////////////////////////////////////////////////////////////////////////////////
+// HostDBInfo methods
+//////////////////////////////////////////////////////////////////////////////////////
+
+HostDBInfo *
+HostDBInfo::alloc(int size)
+{
+  size += sizeof(HostDBInfo);
+  int iobuffer_index = iobuffer_size_to_index(size);
+  ink_release_assert(iobuffer_index >= 0);
+  void *ptr = ioBufAllocator[iobuffer_index].alloc_void();
+  memset(ptr, 0, size);
+  HostDBInfo *ret     = new (ptr) HostDBInfo();
+  ret->iobuffer_index = iobuffer_index;
+  return ret;
+}
+
+void
+HostDBInfo::free()
+{
+  Debug("hostdb", "freeing %d bytes at [%p]", (1 << (7 + iobuffer_index)), this);
+  ioBufAllocator[iobuffer_index].free_void((void *)(this));
+}
+
+HostDBInfo *
+HostDBInfo::unmarshall(char *buf, unsigned int size)
+{
+  if (size < sizeof(HostDBInfo)) {
+    return nullptr;
+  }
+  HostDBInfo *ret = HostDBInfo::alloc(size - sizeof(HostDBInfo));
+  int buf_index   = ret->iobuffer_index;
+  memcpy((void *)ret, buf, size);
+  // Reset the refcount back to 0, this is a bit ugly-- but I'm not sure we want to expose a method
+  // to mess with the refcount, since this is a fairly unique use case
+  ret                 = new (ret) HostDBInfo();
```

Comment:
I'd be more tempted to have a constructor that takes the serialized buffer and not do the `memcpy`. That lets you reset in a constructor, which is much more understandable.